sql/core/v1.2.1/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcFilters.scala

```diff
@@ -153,92 +189,143 @@ private[sql] object OrcFilters extends OrcFiltersBase {
+        filteredSubtree.map(Not(_))
+
+      case EqualTo(attribute, value) if isSearchableType(dataTypeMap(attribute)) => Some(expression)
+      case EqualNullSafe(attribute, value) if isSearchableType(dataTypeMap(attribute)) =>
+        Some(expression)
+      case LessThan(attribute, value) if isSearchableType(dataTypeMap(attribute)) =>
+        Some(expression)
+      case LessThanOrEqual(attribute, value) if isSearchableType(dataTypeMap(attribute)) =>
+        Some(expression)
+      case GreaterThan(attribute, value) if isSearchableType(dataTypeMap(attribute)) =>
+        Some(expression)
+      case GreaterThanOrEqual(attribute, value) if isSearchableType(dataTypeMap(attribute)) =>
+        Some(expression)
+      case IsNull(attribute) if isSearchableType(dataTypeMap(attribute)) => Some(expression)
+      case IsNotNull(attribute) if isSearchableType(dataTypeMap(attribute)) => Some(expression)
+      case In(attribute, values) if isSearchableType(dataTypeMap(attribute)) => Some(expression)
+
+      case _ => None
+    }
+  }
+
+  /**
+   * Build a SearchArgument for a Filter that has already been trimmed so as to only contain
+   * expressions that are convertible to a `SearchArgument`. This allows for a more efficient and
+   * more readable implementation since there's no need to check every node before converting it.
+   *
+   * NOTE: If you change the set of supported `Filter` types here, you need to modify
+   * `trimNonConvertibleSubtreesImpl` accordingly!
+   *
+   * @param dataTypeMap a map from the attribute name to its data type.
+   * @param expression the trimmed input filter predicates.
+   * @param builder the builder so far.
+   * @return
+   */
+  private def createBuilder(
+      dataTypeMap: Map[String, DataType],
+      expression: TrimmedFilter,
+      builder: Builder): Builder = {
+    def getType(attribute: String): PredicateLeaf.Type =
+      getPredicateLeafType(dataTypeMap(attribute))
+
+    import org.apache.spark.sql.sources._
+    def updateBuilder(subexpression: Filter): Unit = subexpression match {
+      case And(left, right) =>
+        builder.startAnd()
+        updateBuilder(left)
+        updateBuilder(right)
+        builder.end()
+
+      case Or(left, right) =>
+        builder.startOr()
+        updateBuilder(left)
+        updateBuilder(right)
+        builder.end()
+
+      case Not(child) =>
+        builder.startNot()
+        updateBuilder(child)
+        builder.end()
 
       // NOTE: For all case branches dealing with leaf predicates below, the additional `startAnd()`
       // call is mandatory.  ORC `SearchArgument` builder requires that all leaf predicates must be
       // wrapped by a "parent" predicate (`And`, `Or`, or `Not`).
 
-      case EqualTo(attribute, value) if isSearchableType(dataTypeMap(attribute)) =>
+      case EqualTo(attribute, value) =>
         val quotedName = quoteAttributeNameIfNeeded(attribute)
```

Comment:
One thing we can do is, add an `assert(isSearchableType(dataTypeMap(attribute)))` at the beginning.
```
case _ =>
  assert(isSearchableType(dataTypeMap(attribute)))
  subexpression match {
    case EqualTo(attribute, value) => ...
    case EqualNullSafe(attribute, value) => ...
    ...
    case _ => throw new IllegalStateException(...)
  }
```

Reply:
This also guarantees that the input `filter` is already trimmed, so that we can get rid of the `TrimmedFilter`.

Reply:
Hm, that's a good idea in principle, but I don't think the outer case-match with the assert can be implemented properly. That is, when I do:

```
      case _ =>
        assert(isSearchableType(dataTypeMap(attribute)))
        subexpression match {
          case EqualTo(attribute, value) =>
      ...
```

, the compiler doesn't know what `attribute` is because the match is simply against `_`. AFAICT there's no common class that I can match against here to say "match any `Filter` that has an `attribute` field" since all the relevant filters are just implemented as separate case classes that just extend Filter directly. The idea of using a structural type to signify "has the field `attribute`" also crossed my mind, but it seems like you can't match on a structural type.

So AFAICT the options for implementing this are:
- Add the assert to every single case. This seems a bit too verbose.
- Bring back the `if isSearchableType(...)` on every case, and then throw an exception in the final `case _` instead of returning true.
What are your thoughts? Another question is whether we actually want this to fail if things go wrong. I'm a bit concerned about failing since it could prevent people from using this if something in the implementation gets inconsistent.

The other thing is that all of this will only happen at runtime. The `TrimmedFilter` is a way to alert the developer at compile time that they shouldn't be calling this method with anything that's not trimmed, whereas these options could fail unexpectedly at runtime.

Let me know what you think about these and we can figure out what to do here.

Reply:
> Add the assert to every single case. This seems a bit too verbose.

I think it's as verbose as adding a `if isSearchableType(...)` for every case and I'm ok with it.

> The TrimmedFilter is a way to alert the developer at compile time that they shouldn't be calling this method with anything that's not trimmed

Personally I think it's just a way to force the developers to look at the doc and see what they should do before calling this method. I don't see a big value of it, as I don't think there will be more callers of this method. It's acceptable to fail some tests and alter the developers to change their code.